pymongo/asynchronous/client_bulk.py

```diff
@@ -133,7 +133,10 @@ def add_insert(self, namespace: str, document: _DocumentOut) -> None:
         validate_is_document_type("document", document)
         # Generate ObjectId client side.
         if not (isinstance(document, RawBSONDocument) or "_id" in document):
-            document["_id"] = ObjectId()
+            new_document = {"_id": ObjectId()}
+            new_document.update(document)
+            document.clear()
```

Comment:
The more I think about it the more I think it's problematic to call clear() and update() here. Those methods could have unintentional side effects aside from the perf problems. For example consider a user passing a custom mapping class which overrides clear()/update().

Reply:
Using ChainMap makes sense, agreed. Explicitly modifying a user-supplied mapping will always carry some risks unfortunately, using the least amount of APIs as possible seems like a safer bet here.

Reply:
Did some more thinking: is the added complexity and changing of the type to `ChainMap` here worth it over this much simpler approach:

```
  if "_id" in document:
      document = {"_id": document["_id"]} | document
  else:
      id = ObjectId()
      document["_id"] = id
      document = {"_id": id} | document
```

If the original document already had an `_id` field, this doesn't modify it, which is consistent with our other insert code paths. If we generated an `_id` field, we still add it to the original document, but we don't worry about the order of the original document.

Reply:
Simpler yes, but it's not performant:
```
$ python -m timeit -s 'd={str(k):k for k in range(10000)};from collections import ChainMap' 'ChainMap(d,{"_id":1})'
2000000 loops, best of 5: 163 nsec per loop
$ python -m timeit -s 'd={str(k):k for k in range(10000)}' '{"_id":1}|d'
2000 loops, best of 5: 143 usec per loop
```

Reply:
Excellent point!

Reply:
Yeah unwrapping it back into the original map makes sense.